File: ios/Classes/model/map/overlay/overlay/NInfoWindow.swift
import NMapsMap

internal struct NInfoWindow: AddableOverlay {
    typealias OverlayType = NMFInfoWindow
    var overlayPayload: Dictionary<String, Any?> = [:]

    let info: NOverlayInfo
    let text: String
    let anchor: NPoint
    let alpha: CGFloat
    let position: NMGLatLng?
    let offsetX: Double
    let offsetY: Double

    func createMapOverlay() -> OverlayType {
        let infoWindow = NMFInfoWindow()
        infoWindow.dataSource = NInfoWindow.createTextSource(text);
        infoWindow.anchor = anchor.cgPoint
        infoWindow.alpha = alpha
        infoWindow.offsetX = asRoundInt(rawFloat: offsetX)
        infoWindow.offsetY = asRoundInt(rawFloat: offsetY)
        if let position = position {
            infoWindow.position = position
        }
        return infoWindow
    }

    static func fromMessageable(_ v: Any) -> NInfoWindow {
        let d = asDict(v)
        return NInfoWindow(
                info: NOverlayInfo.fromMessageable(d[infoName]!),
                text: asString(d[textName]!),
                anchor: NPoint.fromMessageable(d[anchorName]!),
                alpha: asDouble(d[alphaName]!),
                position: castOrNull(d[positionName], caster: asLatLng),
                offsetX: asDouble(d[offsetXName]!),
                offsetY: asDouble(d[offsetYName]!)
        )
    }

    static func createTextSource(_ text: String) -> NMFInfoWindowDefaultTextSource {
        let textSource = NMFInfoWindowDefaultTextSource.data()
        textSource.title = text
        return textSource
    }

    /*
        --- Messaging Name Define ---
    */
    private static let infoName = "info"
    static let textName = "text"
    static let anchorName = "anchor"
    static let alphaName = "alpha"
    static let positionName = "position"
    static let offsetXName = "offsetX"
    static let offsetYName = "offsetY"
    static let closeName = "close"
}

File: android/src/main/kotlin/dev/note11/flutter_naver_map/flutter_naver_map/model/map/overlay/overlay/NInfoWindow.kt
package dev.note11.flutter_naver_map.flutter_naver_map.model.map.overlay.overlay

import android.content.Context
import com.naver.maps.geometry.LatLng
import com.naver.maps.map.overlay.InfoWindow
import dev.note11.flutter_naver_map.flutter_naver_map.converter.DefaultTypeConverter.asDouble
import dev.note11.flutter_naver_map.flutter_naver_map.converter.DefaultTypeConverter.asMap
import dev.note11.flutter_naver_map.flutter_naver_map.converter.MapTypeConverter.asLatLng
import dev.note11.flutter_naver_map.flutter_naver_map.model.base.NPoint
import dev.note11.flutter_naver_map.flutter_naver_map.model.map.info.NOverlayInfo
import dev.note11.flutter_naver_map.flutter_naver_map.util.DisplayUtil.dpToPx

internal data class NInfoWindow(
    override val info: NOverlayInfo,
    val text: String,
    val anchor: NPoint,
    val alpha: Double,
    val position: LatLng?,
    val offsetXDp: Double,
    val offsetYDp: Double,
    private val context: Context?,
) : AddableOverlay<InfoWindow>() {


    override fun createMapOverlay(): InfoWindow = applyAtRawOverlay(InfoWindow())

    override fun applyAtRawOverlay(overlay: InfoWindow) = overlay.also { m ->
        m.adapter = createTextAdapter(text, context!!)
        m.anchor = anchor.toPointF()
        m.alpha = alpha.toFloat()
        if (position != null) m.position = position
        m.offsetX = dpToPx(offsetXDp)
        m.offsetY = dpToPx(offsetYDp)
    }

    companion object {
        fun fromMessageable(rawMap: Any, context: Context): NInfoWindow = rawMap.asMap().let {
            NInfoWindow(
                info = NOverlayInfo.fromMessageable(it[infoName]!!),
                text = it[textName]!!.toString(),
                anchor = NPoint.fromMessageable(it[anchorName]!!),
                alpha = it[alphaName]!!.asDouble(),
                position = it[positionName]?.asLatLng(),
                offsetXDp = it[offsetXName]!!.asDouble(),
                offsetYDp = it[offsetYName]!!.asDouble(),
                context = context,
            )
        }

        fun createTextAdapter(text: String, context: Context): InfoWindow.Adapter =
            object : InfoWindow.DefaultTextAdapter(context) {
                override fun getText(infoWindow: InfoWindow): CharSequence = text
            }

        /*
            --- Messaging Name Define ---
        */

        private const val infoName = "info"
        const val textName = "text"
        const val anchorName = "anchor"
        const val alphaName = "alpha"
        const val positionName = "position"
        const val offsetXName = "offsetX"
        const val offsetYName = "offsetY"
        const val closeName = "close"
    }
}